File: JavaAlgorithm/src/main/java/algo/tzashinorpu/Daily/COM/sumRootToLeaf_1022.java
package algo.tzashinorpu.Daily.COM;

public class sumRootToLeaf_1022 {
    private int res = 0;

    public int sumRootToLeaf(TreeNode root) {
        dfs(root, 0);
        return this.res;
    }

    private void dfs(TreeNode root, int temp) {
        if (root.left == null && root.right == null) {
            this.res += (temp << 1 | root.val);
            return;
        }
        if (root.left != null) {
            dfs(root.left, temp << 1 | root.val);
        }
        if (root.right != null) {
            dfs(root.right, temp << 1 | root.val);
        }
    }
}

File: Clion/ZeroTrac/Tree/sumRootToLeaf_1022.cpp
//
// Created by TzashiNorpu on 10/24/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::sumRootToLeaf(TreeNode *root) {
  int res = 0, tmp = 0;
  function<void(TreeNode *)> dfs = [&](TreeNode *node) -> void {
      if (!node) {
        return;
      }
      if (node->left == nullptr && node->right == nullptr) {
        res += tmp << 1 | node->val;
        return;
      }
      tmp = tmp << 1 | node->val;
      dfs(node->left);
      dfs(node->right);
      tmp = tmp >> 1;
  };
  dfs(root);
  return res;
}
